### PR TITLE
feat(iip-59): protocol-native voter reward distribution

### PR DESCRIPTION
**Pull request by raullenchai**
## Summary

[IIP-59](https://github.com/iotexproject/iips/pull/68): Replace the centralized Hermes reward distribution service with **protocol-native, automatic voter reward distribution**. Zero gas, zero external dependencies, fully trustless.

Companion proto PR: https://github.com/iotexproject/iotex-proto/pull/168

## Problem: Hermes is a Centralized Bottleneck

```
Protocol ──(epoch reward)──► Delegate claims ──► Hermes (centralized) ──► MultiSend ──► Voters
                                                  ├─ Holds private keys
                                                  ├─ Queries external GraphQL
                                                  ├─ Charges service fees
                                                  └─ Single point of failure (unmaintained since 2022)
```

## Solution: Protocol-Native Distribution

```
Protocol ──(epoch reward)──► GrantEpochReward() ──┬──► Delegate (commission)
                                                   ├──► Voter₁ (proportional share)
                                                   ├──► Voter₂ (proportional share)
                                                   └──► Voter_N (proportional share)
                                                         └─ Voters claim via existing ClaimFromRewardingFund
```

## Architecture

```
┌─────────────────────────────────────────────────────────────────────┐
│                     GrantEpochReward() — per epoch                   │
│                                                                     │
│  For each delegate with CommissionRate > 0:                         │
│                                                                     │
│  1. commission = totalReward × CommissionRate / 10000               │
│     └─ grantToAccount(delegate.Reward, commission)                  │
│                                                                     │
│  2. ActiveBucketsByCandidate(sm, candidate)                         │
│     ├─ Native buckets: read _candIndex → NativeBucket(idx)          │
│     └─ Contract staking V1/V2/V3: BucketsByCandidate(addr, height) │
│     └─ Filter: skip isUnstaked() buckets                            │
│                                                                     │
│  3. CalculateVoteWeight(stakingProtocol.VoteWeightCalConsts, ...)   │
│     └─ Aggregate by voter address (not per-bucket)                  │
│                                                                     │
│  4. voterShare = voterPool × weight / totalWeight                   │
│     └─ grantToAccount(voter, voterShare)                            │
│     └─ Rounding dust → delegate                                     │
│                                                                     │
│  CommissionRate == 0 → exact legacy behavior (no change)            │
└─────────────────────────────────────────────────────────────────────┘
```

## Production Safeguards

| Safeguard | Implementation |
|-----------|---------------|
| **Hard fork gate** | `YosemiteBlockHeight` in genesis + `EnableVoterRewardDistribution` feature flag |
| **Commission rate cooldown** | 168-epoch (~7 day) cooldown between changes, enforced on-chain |
| **Contract staking support** | Reads from all 3 contract staking indexers (V1/V2/V3) + native buckets |
| **Backward compatible** | `CommissionRate = 0` (default) = exact legacy behavior, zero impact |
| **Per-voter aggregation** | Multiple buckets aggregated before division, minimizes rounding loss |
| **Consistent vote weights** | Uses `stakingProtocol.VoteWeightCalConsts()`, not hardcoded genesis |
| **Candidate alignment** | `splitEpochReward()` returns filtered list, no misalignment from exempt filtering |
| **Distinct reward log types** | `VOTER_REWARD` (5) and `COMMISSION_REWARD` (6) for indexer/explorer differentiation |

## File Changes

| File | Change |
|------|--------|
| `genesis/genesis.go` | `YosemiteBlockHeight` + `IsYosemite()` |
| `protocol/context.go` | `EnableVoterRewardDistribution` feature flag |
| `staking/stakingpb/staking.proto` | `commissionRate` (11) + `commissionRateLastEpoch` (12) |
| `staking/stakingpb/staking.pb.go` | Generated fields + getters |
| `staking/candidate.go` | `CommissionRate` + `CommissionRateLastEpoch` in struct/Clone/Proto |
| **`staking/voter_reward.go`** (new) | `ActiveBucketsByCandidate()` (native+contract), `CandidateByIdentifier()`, `VoteWeightCalConsts()`, `handleSetCommissionRate()` with cooldown |
| **`action/set_commission_rate.go`** (new) | `SetCommissionRate` action type |
| `staking/protocol.go` | Register handler in dispatch |
| **`rewarding/voter_reward.go`** (new) | `distributeVoterReward()` core logic with feature gate |
| `rewarding/reward.go` | Hook into `GrantEpochReward()`, fix `splitEpochReward` return |
| `rewarding/rewardingpb/rewarding.proto` | `VOTER_REWARD` (5) + `COMMISSION_REWARD` (6) log types |
| **`rewarding/voter_reward_test.go`** (new) | Unit tests: voter shares, commission edge cases, 2800-voter precision |
| **`rewarding/voter_reward_fullscale_test.go`** (new) | Full-scale 24-128 delegate dry run + benchmark |

### Unit Tests — 19/19 PASS

```
TestAdminPb                    PASS    TestProtocol_SetEpochReward     PASS
TestProtocol_Fund              PASS    TestDepositNegativeGasFee       PASS
TestValidateExtension          PASS    TestProtocol_Validate           PASS
TestProtocol_Handle            PASS    TestStateCheckLegacy            PASS
TestMigrateValue               PASS    TestProtocol_GrantBlockReward   PASS
TestProtocol_GrantEpochReward  PASS    TestProtocol_ClaimReward        PASS
TestProtocol_NoRewardAddr      PASS    TestRewardLogCompatibility      PASS
TestProtocol_CalculateReward   PASS    TestCalculateVoterShares        PASS
TestCommissionRateEdgeCases    PASS    TestRoundingPrecision2800Voters PASS
TestFullScale24Delegates       PASS
```

### E2E Tests (Mac Mini + Mac Studio) — All PASS

| Test | Result | Coverage |
|------|--------|----------|
| TestBlockReward | PASS (19s) | Standalone node, RollDPoS consensus, block production + reward |
| TestNativeStaking | PASS (1.5s) | 10 subtests: register, stake, deposit, transfer ownership |
| TestNativeStakingVoteBug | PASS (0.5s) | Candidate register + create stake + deposit to stake |
| TestContractStaking | PASS (5.7s) | NFT staking: lock/unlock/unstake/withdraw/merge/expand/delegate change |
| TestContractStakingV2 | PASS (15.6s) | 26 subtests: batch stake, merge, expand, donate, migrate to trie |
| TestBroadcastNodeInfo | PASS (1.3s) | P2P node info broadcast between nodes |
| TestCreateBlockchain | PASS (0.01s) | Blockchain creation integrity |

### Scale-Up Performance (Mac Studio M3 Ultra, 28-core, 256GB)

| Delegates | Total Voters | Time | Per-Delegate | Memory | Dust (Rau) |
|-----------|-------------|------|--------------|--------|------------|
| **24** | 55,374 | **7.3ms** | 306μs | 6 MB | 27,642 |
| 36 | 86,649 | 11.1ms | 310μs | 9 MB | 43,309 |
| **48** | 112,084 | **14.1ms** | 294μs | 10 MB | 56,034 |
| 64 | 151,237 | 20.8ms | 325μs | 18 MB | 75,756 |
| **96** | 219,623 | **28.7ms** | 299μs | 24 MB | 109,774 |
| 128 | 283,488 | 34.8ms | 272μs | 29 MB | 142,133 |

**Key findings:**
- **Linear scaling**: per-delegate cost is constant ~300μs regardless of delegate count
- **128 delegates × 283K voters = 35ms** — 0.7% of 5-second block time
- **Memory**: 29 MB for 128 delegates — negligible
- **Precision**: dust < voter count (guaranteed by integer division), total exactly conserved

### Benchmark (5-second sustained)

```
BenchmarkDistribute2800Voters-28     16,473 iterations    353μs/op    359KB/op
BenchmarkFullScale24Delegates-28        846 iterations    7.1ms/op    7.1MB/op
```

### Blocked on [iotex-proto PR #168](https://github.com/iotexproject/iotex-proto/pull/168)
- [ ] `FillAction()` / `Proto()` / `LoadProto()` for SetCommissionRate (field 54)
- [ ] Deserializer case in `envelope.go`
- [ ] EthABI encoding for MetaMask/Hardhat
- [ ] `toIoTeXTypes()` expose CommissionRate in CandidateV2 RPC responses
- [ ] `ioctl stake2 setcommission <bps>` CLI command

### Post-merge iteration
- [ ] Commission rate min/max governance limits
- [ ] Voter notification (explorer/wallet integration for rate change alerts)
- [ ] Migration dashboard (delegate opt-in monitoring)
- [ ] Hermes deprecation (archive repos after full migration)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: action/protocol/staking/voter_reward.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+) (*receiptLog, error) {
+	actCtx := protocol.MustGetActionCtx(ctx)
+	blkCtx := protocol.MustGetBlockCtx(ctx)
+	featureCtx := protocol.MustGetFeatureCtx(ctx)
+	log := newReceiptLog(p.addr.String(), HandleSetCommissionRate, featureCtx.NewStakingReceiptFormat)
+
+	// Gate behind Yosemite hard fork
+	if !featureCtx.EnableVoterRewardDistribution {
+		return log, errors.New("SetCommissionRate not enabled before Yosemite hard fork")
+	}
```
Comment (envestcc) [inline action/protocol/staking/voter_reward.go:122]: The check for whether the feature is enabled should be moved to the protocol validation step. This way, the transaction will be ignored during minting rather than causing the mint to fail.

File: action/protocol/staking/voter_reward.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+	// Enforce cooldown: must wait CommissionRateCooldownEpochs between changes
+	rp := rolldpos.MustGetProtocol(protocol.MustGetRegistry(ctx))
+	currentEpoch := rp.GetEpochNum(blkCtx.BlockHeight)
+	if c.CommissionRateLastEpoch > 0 && currentEpoch < c.CommissionRateLastEpoch+action.CommissionRateCooldownEpochs {
+		return log, errors.Errorf(
+			"commission rate cooldown: last changed at epoch %d, current epoch %d, must wait until epoch %d",
+			c.CommissionRateLastEpoch, currentEpoch, c.CommissionRateLastEpoch+action.CommissionRateCooldownEpochs,
+		)
+	}
```
Comment (envestcc) [inline action/protocol/staking/voter_reward.go:138]: When still within the cooldown period, it is better to return a `handleError` so that the transaction status is marked as failed, rather than causing the mint to fail.

File: action/protocol/staking/voter_reward.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+	c.CommissionRate = act.Rate()
+	c.CommissionRateLastEpoch = currentEpoch
+
+	if err := csm.Upsert(c); err != nil {
+		return log, csmErrorToHandleError(c.GetIdentifier().String(), err)
+	}
+
+	log.AddTopics(c.GetIdentifier().Bytes())
+	log.AddAddress(actCtx.Caller)
```
Comment (envestcc) [inline action/protocol/staking/voter_reward.go:148]: The format of transaction events should be compatible with the contract events format. For reference on the contract interface format, see `action/native_staking_contract_interface.sol`.

File: action/protocol/staking/voter_reward.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+)
+
+const (
+	// HandleSetCommissionRate is the handler name
+	HandleSetCommissionRate = "setCommissionRate"
+)
+
+// ActiveBucketsByCandidate returns all non-unstaked buckets (native + contract staking)
+// for a candidate. Used by the rewarding protocol (IIP-59) to distribute voter rewards.
+func (p *Protocol) ActiveBucketsByCandidate(sm protocol.StateManager, candidateIdentifier address.Address) ([]*VoteBucket, error) {
```
Comment (envestcc) [inline action/protocol/staking/voter_reward.go:27]: **Critical issue**: Reading all buckets is an extremely time-consuming and resource-intensive operation, which may cause mint timeouts. It is recommended to dynamically maintain a view to store the votes of each voter.

**Comment by raullenchai** [comment]
Closing as superseded and stale. This was an early (March 2026) take on protocol-native voter reward distribution; the IIP-59 rewarding work is now being implemented by @envestcc's active series (#4924 rework distributeVoterReward, #4926 fold block reward per-delegate, plus #4911/#4912/#4915/#4922/#4923). This PR hasn't advanced since March and its Quality Gate is failing, so it's superseded — closing to keep the queue clean. Happy to reopen if any piece here is still useful as reference.
